ext/pdo/pdo_sql_parser.re

```diff
@@ -91,12 +74,35 @@ PDO_API int pdo_parse_params(pdo_stmt_t *stmt, zend_string *inquery, zend_string
 	struct pdo_bound_param_data *param;
 	int query_type = PDO_PLACEHOLDER_NONE;
 	struct placeholder *placeholders = NULL, *placetail = NULL, *plc = NULL;
+	int (*scan)(pdo_scanner_t *s);
+	struct custom_quote custom_quote = {NULL, 0};
+
+	scan = stmt->dbh->methods->scanner ? stmt->dbh->methods->scanner : default_scanner;
 
 	s.cur = ZSTR_VAL(inquery);
 	s.end = s.cur + ZSTR_LEN(inquery) + 1;
 
 	/* phase 1: look for args */
 	while((t = scan(&s)) != PDO_PARSER_EOI) {
+		if (custom_quote.pos) {
+			/* Inside a custom quote */
+			if (t == PDO_PARSER_CUSTOM_QUOTE && custom_quote.len == s.cur - s.tok && !strncmp(s.tok, custom_quote.pos, custom_quote.len)) {
+				/* Matching closing quote found, end custom quoting */
+				custom_quote.pos = NULL;
+				custom_quote.len = 0;
+			}
+
+			continue;
```

Comment:
I absolutely do not want to slow this PR that I find a wonderful and clean improvement for pg users,
but wouldn't prudence dictate to _not_ continue in case of a PDO_PARSER_ESCAPED_QUESTION, but instead handle it and emit a deprecation notice here until a next major version?

This would be for the corner case of those having relied on your [PHP RFC: Escape PDO "?" parameter placeholder](https://wiki.php.net/rfc/pdo_escape_placeholders) to pass question marks through dollar-quoted strings:
there this new PR could be seen as a BC break.
(summary in [https://github.com/php/php-src/issues/14244#issuecomment-2112635487] you commented on)

I can't say if anybody uses it (I for one relied on ATTR_EMULATE_PREPARES), but just in case…

Reply:
Thanks for the heads up, I'll look into it!

Reply:
This should be taken care in [16a26d4](https://github.com/php/php-src/pull/14035/commits/16a26d4fbdac3721b1a69bcd8f61b05f5f70e723)